From 864cd2f607f9d1b1831d9c73e17baf530484da37 Mon Sep 17 00:00:00 2001
From: Petri Hintukainen <phintuka@gmail.com>
Date: Wed, 5 Aug 2026 18:36:03 +0300
Subject: [PATCH] graphics_processor: Fix OOB write on invalid input

Too small buffer size resulted in integer overflow when calculating
size of segment data.

Reported-by: Dominik Riedweg <domi.riedweg@gmail.com>
---
 src/libbluray/decoders/graphics_processor.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/libbluray/decoders/graphics_processor.c b/src/libbluray/decoders/graphics_processor.c
index 3638a6da..cb431149 100644
--- a/src/libbluray/decoders/graphics_processor.c
+++ b/src/libbluray/decoders/graphics_processor.c
@@ -179,7 +179,9 @@ static int _join_segment_fragments(struct pes_buffer_s *p)
         bb_init(&bb, next->buf + sd_pos, 3);
         pg_decode_sequence_descriptor(&bb, &sd);
 
-        _join_fragments(p, next, data_pos);
+        if (next->len > data_pos /* at least one byte of payload */) {
+            _join_fragments(p, next, data_pos);
+        }
 
         pes_buffer_remove(&p, next);
 
